src/code/padmgr.c

```diff
@@ -1,58 +1,797 @@
+                            }
+                            triedRumbleComm = true;
+                        }
+                    } else {
+                        if (sPadMgrInstance->rumbleTimer[i] != 0) {
+                            // Rumble Pak stop
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/padmgr/Padmgr_Update.s")
+                            if (osMotorStop(&sPadMgrInstance->rumblePfs[i]) != 0) {
+                                // error
+                                sPadMgrInstance->pakType[i] = CONT_PAK_NONE;
+                            } else {
+                                sPadMgrInstance->rumbleTimer[i]--;
+                            }
+                            triedRumbleComm = true;
+                        }
+                    }
+                }
+            } else if (sPadMgrInstance->pakType[i] != CONT_PAK_NONE) {
+                sPadMgrInstance->pakType[i] = CONT_PAK_NONE;
+            }
+        }
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/padmgr/Padmgr_Stop.s")
+    if (!triedRumbleComm) {
+        // Try to initialize the rumble pak for controller port `i` if a controller pak is connected and not already
+        // known to be an initialized a rumble pak
+        i = sPadMgrRetraceCount % MAXCONTROLLERS;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/padmgr/func_80175E68.s")
+        if (sPadMgrInstance->ctrlrType[i] == PADMGR_CONT_NORMAL &&
+            (sPadMgrInstance->padStatus[i].status & CONT_CARD_ON) && sPadMgrInstance->pakType[i] != CONT_PAK_RUMBLE) {
+            ret = osMotorInit(serialEventQueue, &sPadMgrInstance->rumblePfs[i], i);
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/padmgr/Padmgr_GetInput.s")
+            if (ret == 0) {
+                // Got rumble pak
+                sPadMgrInstance->pakType[i] = CONT_PAK_RUMBLE;
+                sPadMgrInstance->rumbleTimer[i] = 2;
+            } else if (ret == PFS_ERR_DEVICE) {
+                // Not a rumble pak
+                if (sPadMgrInstance->pakType[i] != CONT_PAK_OTHER) {
+                    sPadMgrInstance->pakType[i] = CONT_PAK_OTHER;
+                }
+            } else if (ret == PFS_ERR_CONTRFAIL) {
+                // Pak communication error
+            } else if (ret == PFS_ERR_NOPACK) {
+                // No controller pak
+            } else {
+                // Unrecognized return code
+                Fault_AddHungupAndCrash("../padmgr.c", 594);
+            }
+        }
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/padmgr/Padmgr_GetInput2.s")
+    PadMgr_ReleaseSerialEventQueue(serialEventQueue);
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/padmgr/Padmgr_ThreadEntry.s")
+/**
+ * Immediately stops rumble on all controllers
+ */
+void PadMgr_RumbleStop(void) {
+    OSMesgQueue* serialEventQueue = PadMgr_AcquireSerialEventQueue();
+    s32 i;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/padmgr/PadMgr_Init.s")
+    for (i = 0; i != MAXCONTROLLERS; i++) {
```

Comment:
```suggestion
    for (i = 0; i < MAXCONTROLLERS; i++) {
```
Does this still match?